web/src/main/java/org/springframework/security/web/server/authentication/SwitchUserWebFilter.java

```diff
@@ -0,0 +1,366 @@
+
+		final List<GrantedAuthority> extendedTargetUserAuthorities = new ArrayList<>(targetUserAuthorities);
+		extendedTargetUserAuthorities.add(switchAuthority);
+
+		return new UsernamePasswordAuthenticationToken(
+				targetUser, targetUser.getPassword(), extendedTargetUserAuthorities
+		);
+	}
+
+	/**
+	 * Find the original <code>Authentication</code> object from the current user's
+	 * granted authorities. A successfully switched user should have a
+	 * <code>SwitchUserGrantedAuthority</code> that contains the original source user
+	 * <code>Authentication</code> object.
+	 *
+	 * @param currentAuthentication The current <code>Authentication</code> object
+	 * @return The source user <code>Authentication</code> object or <code>Optional.empty</code>
+	 * otherwise.
+	 */
+	private Optional<Authentication> extractSourceAuthentication(Authentication currentAuthentication) {
+		// iterate over granted authorities and find the 'switch user' authority
+		return currentAuthentication.getAuthorities().stream()
+				.filter(a -> a instanceof SwitchUserGrantedAuthority)
+				.map(a -> (SwitchUserGrantedAuthority) a)
+				.map(SwitchUserGrantedAuthority::getSource)
+				.findFirst();
+	}
+
+	private static ServerWebExchangeMatcher createMatcher(String pattern) {
+		return ServerWebExchangeMatchers.pathMatchers(HttpMethod.POST, pattern);
+	}
+
+	private AuthenticationCredentialsNotFoundException noCurrentUserException() {
+		return new AuthenticationCredentialsNotFoundException(
+				"No current user associated with this request"
+		);
+	}
+
+	private AuthenticationCredentialsNotFoundException noOriginalAuthenticationException() {
+		return new AuthenticationCredentialsNotFoundException(
+				"Could not find original Authentication object"
+		);
+	}
+
+	private AuthenticationCredentialsNotFoundException noTargetAuthenticationException() {
+		return new AuthenticationCredentialsNotFoundException(
+				"No target user for the given username"
+		);
+	}
+
+	private static class SwitchUserAuthenticationException extends RuntimeException {
+		SwitchUserAuthenticationException(AuthenticationException exception) {
+			super(exception);
+		}
+	}
+
+	/**
+	 * Sets the repository for persisting the SecurityContext. Default is {@link WebSessionServerSecurityContextRepository}
+	 *
+	 * @param securityContextRepository the repository to use
+	 */
+	public void setSecurityContextRepository(
+			ServerSecurityContextRepository securityContextRepository) {
+		Assert.notNull(securityContextRepository, "securityContextRepository cannot be null");
+		this.securityContextRepository = securityContextRepository;
+	}
+
+	/**
+	 * Sets the userDetails object validator
+	 *
+	 * @param userDetailsChecker The <tt>userDetailsChecker</tt> which will be used to check
+	 *                           the data for the user that is being switched to.
+	 */
+	public void setUserDetailsChecker(UserDetailsChecker userDetailsChecker) {
+		this.userDetailsChecker = userDetailsChecker;
+	}
+
```

Comment:
All of the public methods below should have tests that ensure they are working as expected.

Reply:
done